group-coordinator/src/main/java/org/apache/kafka/coordinator/group/generic/GenericGroup.java

```diff
@@ -898,6 +904,53 @@ public void createGroupTombstoneRecords(List<Record> records) {
         records.add(RecordHelpers.newGroupMetadataTombstoneRecord(groupId()));
     }
 
+    @Override
+    public boolean isEligibleForDeletion() {
+        return isInState(EMPTY) && generationId > 0;
+    }
+
+    /**
+     * Return the expiration condition to be used for this group. This is based on several factors
+     * such as the group state, the protocol type, and the GroupMetadata record version.
+     * See {@link org.apache.kafka.coordinator.group.OffsetMetadataManager.ExpirationCondition}
+     *
+     * @return The expiration condition.
+     */
+    @Override
+    public OffsetMetadataManager.ExpirationCondition expirationCondition() {
+        if (protocolType.isPresent()) {
+            if (isInState(EMPTY)) {
+                // No consumer exists in the group =>
+                // - If current state timestamp exists and retention period has passed since group became Empty,
+                //   expire all offsets with no pending offset commit;
+                // - If there is no current state timestamp (old group metadata schema) and retention period has passed
+                //   since the last commit timestamp, expire the offset
+                return new OffsetMetadataManager.ExpirationCondition(
+                    offsetAndMetadata -> currentStateTimestamp.orElse(offsetAndMetadata.commitTimestampMs),
+                    Collections.emptySet());
+            } else if (ConsumerProtocol.PROTOCOL_TYPE.equals(protocolType.get()) &&
+                subscribedTopics.isPresent() && isInState(STABLE)) {
```

Comment:
nit: Could we align subscribedTopics with ConsumerProtocol on the previous line?